krm-functions/lib/nad/v1/nad.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return nil, err
+	}
+	return &NadStruct{K: *p}, nil
+}
+
+func (r *NadStruct) GetCniSpecType(fields ...string) CniSpecType {
+	return r.C
+}
+
+func (r *NadStruct) GetStringValue(fields ...string) string {
+	if r == nil {
+		return ""
+	}
+	s, ok, err := r.K.NestedString(fields...)
+	if err != nil {
+		return ""
+	}
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func (r *NadStruct) GetConfigSpec() string {
+	if r == nil {
+		return ""
+	}
+	s, ok, err := r.K.NestedString(ConfigType...)
+	if err != nil {
+		return ""
+	}
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func (r *NadStruct) GetCNIType() string {
+	nadConfigStruct := NadConfig{}
+	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
+		return ""
+	}
+	return nadConfigStruct.Plugins[0].Type
+}
+
+func (r *NadStruct) GetVlan() int {
+	nadConfigStruct := NadConfig{}
+	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
+		return 0
+	}
+	return nadConfigStruct.Vlan
+}
+
+func (r *NadStruct) GetNadMaster() string {
+	nadConfigStruct := NadConfig{}
+	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
+		return ""
+	}
+	return nadConfigStruct.Plugins[0].Master
+}
+
+func (r *NadStruct) GetIpamAddress() []Addresses {
+	nadConfigStruct := NadConfig{}
+	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
+		return []Addresses{}
+	}
+	return nadConfigStruct.Plugins[0].Ipam.Addresses
+}
+
+func (r *NadStruct) GetNadConfig() NadConfig {
+	nadConfigStruct := NadConfig{}
+	configSpec := r.GetStringValue(ConfigType...)
+	if configSpec == "" {
+		configSpec = "{}"
+	}
+	if err := json.Unmarshal([]byte(configSpec), &nadConfigStruct); err != nil {
+		panic(err)
+	}
+	nadConfigStruct.CniVersion = CniVersion
+	if r.GetCniSpecType() == VlanType {
+		return nadConfigStruct
+	} else if r.GetCniSpecType() == IpamType || r.GetCniSpecType() == Both {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
+						Ips: true,
+					},
+					Mode: NadMode,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct
+	}
+	return nadConfigStruct
+}
+
+func (r *NadStruct) SetCniSpecType(cniType CniSpecType) {
+	r.C = cniType
+}
+
+// SetConfigSpec sets the spec attributes in the kubeobject according the go struct
+func (r *NadStruct) SetConfigSpec(spec *nadv1.NetworkAttachmentDefinitionSpec) error {
+	b, err := json.Marshal(spec.Config)
+	if err != nil {
+		return err
+	}
+	return r.K.SetNestedString(string(b), ConfigType...)
+}
+
+func (r *NadStruct) SetCNIType(cniType string) error {
+	if cniType != "" {
+		nadConfigStruct := r.GetNadConfig()
+		nadConfigStruct.Plugins[0].Type = cniType
```

Comment:
check if `len(.....Plugins) > 0` before getting the first element

Reply:
If GetNadConfig() would guarantee that Plugins is not empty this would be fine (I would still write this fact a comment before this line), but right now it doesn't give you that guarantee (and probably it is good this way, and it shouldn't)

Reply:
commented and refactored